clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+// RUN: %clang_cc1 %s --std=c++17 -triple x86_64-pc-windows-msvc -fsycl-is-device -verify -fsyntax-only -Wno-unused
+// RUN: %clang_cc1 %s --std=c++17 -triple x86_64-linux-gnu -fsycl-is-device -verify -fsyntax-only -Wno-unused
+
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) { // #kernelSingleTask
+  kernelFunc();
+}
+
+// kernel3 - expect error
+// The current function is named with a lambda (i.e., takes a lambda as a
+// template parameter. Call the builtin on the current function
+// Then it is passed to kernel
+// Test that passing the current function to the unique stable name builtin and
+// then to the kernel throws an error because the latter causes its name mangling
+// to change
+template <typename Func>
+void kernel3func(const Func &F3) {
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3); // #USN_F3
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#kernel3func_call{{in instantiation of function template specialization}}
+  // expected-note@#USN_F3{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel3>(F3); // #kernel3_call
+}
+
+void callkernel3() {
+  kernel3func([]() {}); // #kernel3func_call
+}
+
+// kernel4 - expect error
+// The current function is named with a lambda (i.e., takes a lambda as a
+// template parameter. Call the builtin on the current function
+// Then an empty lambda is passed to kernel
+// Test that passing the current function to the unique stable name builtin and
+// then passing a lambda to the kernel throws an error because the current function
+// is part of naming the kernel although it is not passed to the kernel, and its
+// mangling changes due to kernel call with the lambda
```

Comment:
```suggestion
// The current function is named with a lambda (i.e., takes a lambda as a
// template parameter). Call the builtin on the given function,
// then an empty lambda is passed to kernel.
// Test that passing the given function to the unique stable name builtin and
// then passing a different lambda to the kernel still throws an error because the calling
// context is part of naming the kernel. Even though the given function (F4) is not passed to the kernel, its
// mangling changes due to kernel call with the unrelated lambda.
```